### PR TITLE
Convert from Cake to Nuke

### PR DESCRIPTION
**Pull request by Silcoish**
Part of the effort to migrate from Cake -> Nuke. I used the conversion tool created for us by [@Matkoch](https://github.com/matkoch) which can take a Cake file and spit out a decent amount of the Build.cs file. 

I've compared the outputs of the files compared to the last build on master and they are almost identical in size, differing in places where you'd expect (eg. the nuspec file as it has a longer version name).

### CODE REVIEW

File: build/Build.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                .SetVersion(OctoVersionInfo.FullSemVer));
+        });
+
+    Target Compile => _ => _
+        .DependsOn(Restore)
+        .DependsOn(Clean)
+        .Executes(() =>
+    {
+        DotNetBuild(_ => _
+            .SetProjectFile(RootDirectory / "source")
```
Comment (matt-richardson) [inline build/Build.cs:64]: Consider pulling this up to a named directory - it's used in multiple spots

File: build/Build.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            .SetConfiguration(Configuration)
+            .SetVersion(OctoVersionInfo.FullSemVer));
+    });
+
+    Target Test => _ => _
+        .DependsOn(Compile)
+        .Executes(() =>
+    {
+        (RootDirectory).GlobFiles("**/**/*.Tests.csproj").ToList().ForEach(testProjectFile =>
```
Comment (matt-richardson) [inline build/Build.cs:73]: does this work?
```suggestion
        RootDirectory.GlobFiles("**/**/*.Tests.csproj").ForEach(testProjectFile =>
```

File: build/Build.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        {
+            DotNetTest(_ => _
+            .SetProjectFile(testProjectFile)
```
Comment (matt-richardson) [inline build/Build.cs:76]: nit: indenting

File: build/Build.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            .SetConfiguration(Configuration)
+            .SetNoBuild(true));
+        });
+    });
+
+    Target Merge => _ => _
+        .DependsOn(Test)
+        .Executes(() =>
+        {
+        var inputFolder = OctopusClientFolder / "bin" / Configuration / "net452";
```
Comment (matt-richardson) [inline build/Build.cs:86]: nit: indenting

File: build/Build.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        var outputFolder = OctopusClientFolder / "bin" / Configuration / "net452Merged";
+        EnsureExistingDirectory(outputFolder);
+        
+        var assemblyPaths = System.IO.Directory.EnumerateFiles(inputFolder, "NewtonSoft.Json.dll").Select(f => (AbsolutePath)f);
```
Comment (matt-richardson) [inline build/Build.cs:90]: Can we just glob here instead? Seems like we're mixing and matching between nuke and native dotnet

File: build/Build.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        assemblyPaths = assemblyPaths.Concat(System.IO.Directory.EnumerateFiles(inputFolder, "Octodiff.exe").Select(f => (AbsolutePath)f));
+        
+        var inputAssemblies = new List<string> { $"{inputFolder}/Octopus.Client.dll" };
```
Comment (matt-richardson) [inline build/Build.cs:93]:
```suggestion
        var inputAssemblies = new List<string> { inputFolder / "Octopus.Client.dll" };
```

File: build/_build.csproj [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    <RootNamespace></RootNamespace>
+    <NoWarn>CS0649;CS0169</NoWarn>
+    <NukeRootDirectory>..</NukeRootDirectory>
+    <NukeScriptDirectory>..</NukeScriptDirectory>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="ILRepack.Lib" Version="2.1.0-beta1" />
+    <PackageReference Include="Nuke.Common" Version="5.0.2" />
+    <PackageDownload Include="GitVersion.Tool" Version="[5.5.1]" />
```
Comment (matt-richardson) [inline build/_build.csproj:15]: Not sure we need Gitversion anymore?

**Review by matt-richardson — APPROVED** [review]
Looks pretty good to me. A couple of minor suggestions.
Looks like the build is 💚 
Haven't tested/compared - let me know if you want me to look deeper.

File: build/Build.cs (hunk 5 above)
Comment (Silcoish) [reply build/Build.cs:90]: Good call - I changed: 

```
var assemblyPaths = System.IO.Directory.EnumerateFiles(inputFolder, "NewtonSoft.Json.dll").Select(f => (AbsolutePath)f);
assemblyPaths = assemblyPaths.Concat(System.IO.Directory.EnumerateFiles(inputFolder, "Octodiff.exe").Select(f => (AbsolutePath)f));

var inputAssemblies = new List<string> { $"{inputFolder}/Octopus.Client.dll" };
inputAssemblies.AddRange(assemblyPaths.Select(x => x.ToString()));
```

to

```
var inputAssemblies = 
                inputFolder.GlobFiles("NewtonSoft.Json.dll", "Octodiff.exe", "Octopus.Client.dll");
```

**Pull request merged**